src/Microsoft.OData.Edm/Csdl/Parsing/AnnotationJsonParser.cs

```diff
@@ -0,0 +1,515 @@
+        /// <returns>true/false</returns>
+        private static bool TryParseLabeledElementReferenceExpression(JsonElement element, JsonParserContext context, out CsdlLabeledExpressionReferenceExpression labeledReferenceExp)
+        {
+            Debug.Assert(context != null);
+            labeledReferenceExp = null;
+
+            // Labeled element reference expressions are represented as an object with a member $LabeledElementReference whose value is a string containing an qualified name.
+            JsonElement propertyValue;
+            if (element.ValueKind != JsonValueKind.Object || !element.TryGetProperty("$LabeledElementReference", out propertyValue))
+            {
+                return false;
+            }
+
+            string label = propertyValue.ProcessProperty("$LabeledElementReference", context, (e, c) => e.ParseAsString(c));
+
+            labeledReferenceExp = new CsdlLabeledExpressionReferenceExpression(label, context.Location());
+            return true;
+        }
+
+        /// <summary>
+        /// Parse the record expression expression using <see cref="JsonElement"/>.
+        /// </summary>
+        /// <param name="element">The input JSON element.</param>
+        /// <param name="context">The parser context.</param>
+        /// <returns>the built record expression.</returns>
+        private static CsdlRecordExpression ParseRecordExpression(JsonElement element, JsonParserContext context)
+        {
+            Debug.Assert(context != null);
+
+            // A record expression MAY specify the structured type of its result, which MUST be an entity type or complex type in scope.
+            // If not explicitly specified, the type is derived from the expression�s context
+
+            // The type of a record expression is represented as the @type control information
+            // for example: "@type": "https://example.org/vocabs/person#org.example.person.Manager",
+            // So far, ODL doesn't support the type "@type" with relative path, only supports like "#Model.VipCustomer", or without #
+            // for 4.0, this name MUST be prefixed with the hash symbol (#);
+            // or non-OData 4.0 payloads, built-in primitive type values SHOULD be represented without the hash symbol,
+            // but consumers of 4.01 or greater payloads MUST support values with or without the hash symbol.
+
+            CsdlTypeReference typeReference = null;
+            if (element.TryGetProperty("@type", out JsonElement typeValue))
+            {
+                // Try to build the type. The type should be "Complex type" or "Entity Type".
+                string typeName = typeValue.ProcessProperty("@type", context, (e, c) => e.ParseAsString(c));
+                int index = typeName.IndexOf('#');
+                if (index >= 0)
+                {
+                    typeName = typeName.Substring(index + 1); // remove the "#"
+                }

```

Comment:
line 435 looks for the "#" anywhere in the typeName, but 438 assumes it is the first character.   In 435 can you just check the first character?
also, the # is required in 4.01 for complex types, entity types, enum types, and type definitions -- just not built-in primitive types.

Reply:
Emm, It seems I don's assume "#" is the first character, and I don't assume there must has a "#" in the type name.

My logic here is that:

1) If a type name string read from the JSON has "#", retrieve the type name after the "#",
2) otherwise, use the whole string read from the JSON as the type name.

---
In reply to: [443016823](https://github.com/OData/odata.net/pull/1775#discussion_r443016823) [](ancestors = 443016823)

Reply:
Right; my bad; I misread.

---
In reply to: [444550195](https://github.com/OData/odata.net/pull/1775#discussion_r444550195) [](ancestors = 444550195,443016823)